### PR TITLE
feat: Add selected row support to GridTable.

### PR DESCRIPTION
**Pull request by stephenh**
Leverages the new mobx-based `RowState` to add a map of `rowId` --> `checked | unchecked | partial`.

Uses the `Checkbox` `indeterminate` support to show partially-selected parents as indeterminate.

Handles clicking a parent --> make all children equally selected/unselected.

Handles clicking a child --> recalc it's parents "are you all/none/partially selected", recursively.

### CODE REVIEW

File: src/components/Table/GridTable.test.tsx [hunk 1]
```diff
@@ -913,9 +922,9 @@ describe("GridTable", () => {
     // When we filter by 'foo'
     const r = await render(<GridTable filter="foo" columns={nestedColumns} rows={rows} />);
     // Then we show the header
-    expect(cell(r, 0, 1)).toHaveTextContent("Name");
+    expect(cell(r, 0, 2)).toHaveTextContent("Name");
```
Comment (stephenh) [inline src/components/Table/GridTable.test.tsx:925]: With a new column for testing "select", some existing assertions moved their column over by 1.

File: src/components/Table/GridTable.test.tsx [hunk 2]
```diff
@@ -1036,6 +1045,48 @@ describe("GridTable", () => {
...
+        ...{ kind: "parent", id: "p1", name: "parent 1" },
+        children: [
+          {
+            ...{ kind: "child", id: "p1c1", name: "child p1c1" },
+            children: [{ kind: "grandChild", id: "p1c1g1", name: "grandchild p1c1g1" }],
+          },
+        ],
+      },
+    ];
+    const api: MutableRefObject<GridTableApi | undefined> = { current: undefined };
```
Comment (stephenh) [inline src/components/Table/GridTable.test.tsx:1062]: In production code, this `api` prop would be a `useRef` that the page passes into `GridTable` to "get back" a pointer to API methods it can call to ask/interrogate `GridTable` about various things, i.e. what is your current list of selected rows.

Note that I don't have an `onRowsSelected` callback yet...should be easy to add...will probably do as a separate PR...

File: src/components/Table/GridTable.tsx [hunk 3]
```diff
@@ -959,7 +966,7 @@ function GridRow<R extends Kinded, S>(props: GridRowProps<R, S>): ReactElement {
           .filter((style) => style)
       : undefined;
 
-  const rowStyleCellCss = maybeApplyFunction(row, rowStyle?.cellCss);
+  const rowStyleCellCss = maybeApplyFunction(row as any, rowStyle?.cellCss);
```
Comment (stephenh) [inline src/components/Table/GridTable.tsx:969]: The `IfAny` I did to "fix" `GridDataRow<any>` broke these two other spots, just doing an `as any` on them for now.

File: src/inputs/Checkbox.tsx [hunk 4]
```diff
@@ -6,17 +6,12 @@ import { CheckboxBase } from "src/inputs/CheckboxBase";
 export interface CheckboxProps {
   label: string;
   checkboxOnly?: boolean;
+  selected: boolean | "indeterminate";
```
Comment (stephenh) [inline src/inputs/Checkbox.tsx:9]: So, dunno, I always thought it was kind of weird that "indeterminate" was considered orthogonal to selected, b/c afaiu a checkbox really can be only 1 of "selected yes", "selected no", or "indeterminate".

So I nudged it that way, by removing the `indeterminate` prop and instead having callers pass `selected="indeterminate"` to highlight it doesn't really mean anything to be "selected _and_ indeterminate".

Comment (TylerR909) [reply src/inputs/Checkbox.tsx:9]: Doesn't really sit right with me to have a `boolean | "someString"` type but it doesn't appear that an Enum actually simplifies anything.

File: src/inputs/Checkbox.tsx [hunk 5]
```diff
@@ -26,18 +21,14 @@ export interface CheckboxProps {
...
-    label,
-    indeterminate: isIndeterminate = false,
-    disabled: isDisabled = false,
-    selected,
-    ...otherProps
-  } = props;
-  const ariaProps = { isSelected: selected, isDisabled, isIndeterminate, ...otherProps };
+  const { label, disabled: isDisabled = false, selected, ...otherProps } = props;
+  // Treat indeterminate as false so that clicking on indeterminate always goes --> true.
+  const isSelected = selected === "indeterminate" ? false : selected;
```
Comment (TylerR909) [inline src/inputs/Checkbox.tsx:26]: Same as `selected === true`?

File: src/inputs/Checkbox.stories.tsx [hunk 6]
```diff
@@ -16,14 +16,16 @@ export function Checkboxes() {
...
             onBlur={action("onBlur")}
             label="Default"
           />
-          <Checkbox onChange={action("onChange")} selected label="Selected" />
-          <Checkbox onChange={action("onChange")} indeterminate label="Indeterminate" />
-          <Checkbox onChange={action("onChange")} disabled label="Disabled" />
+          <Checkbox onChange={action("onChange")} selected={false} label="Selected" />
+          <Checkbox onChange={action("onChange")} selected="indeterminate" label="Indeterminate" />
+          <Checkbox onChange={action("onChange")} selected={false} disabled label="Disabled while unselected" />
+          <Checkbox onChange={action("onChange")} selected={true} disabled label="Disabled while selected" />
```
Comment (TylerR909) [inline src/inputs/Checkbox.stories.tsx:28]: Where do we expect `selected` will be coming from? A database field? Manually set as a prop by a developer? Form State?

Just wondering how we programmatically get from some value to `true | false | "indeterminate"`

**Review by DeanGilewicz — APPROVED** [review]
Looks good - I think a nice use of `indeterminate` in this case to represent "partially" selected (some but not all children selected).

File: src/components/Table/RowState.ts [hunk 7]
```diff
@@ -37,6 +35,82 @@ export class RowState {
...
+      // Everything here & down is deterministically on/off
+      const map = new Map<string, SelectedState>();
+      const todo = [curr.row];
+      while (todo.length > 0) {
+        const row = todo.pop()!;
+        map.set(row.id, selected ? "checked" : "unchecked");
+        if (row.children) {
+          todo.push(...row.children);
+        }
+      }
```
Comment (TylerR909) [inline src/components/Table/RowState.ts:94]: Dang. If only we could tell Array.prototype.flat() what key to flatten on this could be so much more succinct. 

```js
const map = new Map<string, SelectedState>();
curr.row.flatIshMap((row) => map.set(row.id, selected ? "checked" : "unchecked"));
```

Not generally a fan of while/for loops as they're more imperative than declarative. But, they work, so can't really fuss too much.

I do feel like a helper function could assist?

```js
function traverseWithChildren(items, fn, parent) {
  items.forEach((item) => {
    fn(item);
    if (item.children) traverseWithChildren(item.children, fn, item);
  });
}

traverseWithChildren(curr.row,
  (item) => map.set(row.id, selected ? "checked" : "unchecked"),
);
```

**Review by TylerR909 — APPROVED** [review]

File: src/components/Table/RowState.ts (hunk 7 above)
Comment (stephenh) [reply src/components/Table/RowState.ts]: Ah nice, good call!

When writing the 2nd `while`, I thought of that, but swear that it was "different from the other `while`" at first; but turns out it either never was, or at least is not anymore.

So I added a `visit` method giving the "visitor pattern" is exactly what this sort of "please traverse this complex (okay, for us just a tree) data structure for me, and just call my function as necessary kthx" pattern is typically called:

```
function visit(rows: GridDataRow<any>[], fn: (row: GridDataRow<any>) => void): void {
  const todo = [...rows];
  while (todo.length > 0) {
    const row = todo.pop()!;
    fn(row);
    if (row.children) {
      todo.push(...row.children);
    }
  }
}
```

Pushed up.

File: src/inputs/Checkbox.stories.tsx (hunk 6 above)
Comment (stephenh) [reply src/inputs/Checkbox.stories.tsx:28]: Yeah, it will nearly always come from a database boolean field of like "is this enabled yes/no".

I.e. I don't think I've ever seen indeterminate used for anything other than "partially selected rows in a table of parent/child selectable rows" :-) so I'm not too worried about how it would show up in other "just a regular CRUD form" places.

File: src/inputs/Checkbox.tsx (hunk 4 above)
Comment (stephenh) [reply src/inputs/Checkbox.tsx:9]: Ah yeah, I admit "mixing separate primitive types in a single union type" is maybe slightly esoteric functional programming sort of a thing, but agreed it's net/net what an enum would be, but generally same (if not more) simplicity, especially for the ~90% case of "it is really just either true/false".

File: src/inputs/Checkbox.tsx (hunk 5 above)
Comment (stephenh) [reply src/inputs/Checkbox.tsx:26]: Yep, will change
